compiler/rustc_type_ir/src/search_graph/mod.rs

```diff
@@ -941,6 +980,22 @@ impl<D: Delegate<Cx = X>, X: Cx> SearchGraph<D> {
                         heads.insert(head_index, PathsToNested::EMPTY, HeadUsages::default());
                     }
                 } else {
+                    // We may need to mutate the result of our cycle head in case we did not reach
+                    // a fixpoint.
+                    *result = match rebase_reason {
+                        RebaseReason::NoCycleUsages => return false,
+                        RebaseReason::Ambiguity => D::propagate_ambiguity(cx, input, *result),
+                        RebaseReason::Overflow => D::on_fixpoint_overflow(cx, input),
+                        RebaseReason::ReachedFixpoint(None) => *result,
+                        RebaseReason::ReachedFixpoint(Some(path_kind)) => {
+                            if popped_head.usages.is_single(path_kind) {
+                                *result
+                            } else {
+                                return false;
+                            }
+                        }
+                    };
```

Comment:
same with mutating the result. as this can now cause us to drop the provisional result, let's do it before recomputing paths